### PR TITLE
Bootstrap EBNF parser

### PR DESCRIPTION
**Pull request by sahejsingh**
Here is what this does

- Parse EBNF Grammar, get AST
- process AST to build EBNF Parser
- Use the EBNF parser to parse EBNF rules passed in as text

@anz-bank/sysl-developers

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -4,6 +4,8 @@ root = true
 end_of_line = lf
 insert_final_newline = true
 trim_trailing_whitespace = true
+indent_style = space
+indent_size = 4
```
Comment (juliaogris) [inline .editorconfig]: Isn't golang tab-based?
I thought go is very specific in required style, but maybe I'm mistaken.

**Comment by juliaogris** [comment]
Just curious: is there a reason why you chose EBNF over ABNF?

File: sysl2/Grammar1_test.go [hunk 2]
```diff
@@ -6,25 +6,29 @@ import (
     sysl "anz-bank/sysl/sysl2/proto"
 )
 
-func testParser(g *sysl.Grammar, numTerms int, tokens []int, text string, expected bool, t *testing.T) {
+func testParser(g *sysl.Grammar, numTerms int, tokens []int, text string, expectedResult bool, t *testing.T) (bool, []interface{}) {
```
Comment (juliaogris) [inline sysl2/Grammar1_test.go]: How about  something like this (pseudo-code):
```
expected = {
    grammarTermsCnt: 6, 
    grammarParsesText: true, 
    textTokens: [2,3,2,-1]
}
testParser(grammar, text, expected)
```
 or at least `grammar` and `text` as the first two params and all other params prefixed with `expected`?

Comment (sahejsingh) [reply sysl2/Grammar1_test.go]: more effort than it is worth at the moment. Will keep this in mind for later...

File: .editorconfig (hunk 1 above)
Comment (sahejsingh) [reply .editorconfig]: its actually whatever gofmt says.
I have removed it for now.

File: sysl2/Grammar1_test.go (hunk 2 above)
Comment (juliaogris) [reply sysl2/Grammar1_test.go]: kk

File: sysl2/Grammar1_test.go [hunk 3]
```diff
@@ -251,3 +248,68 @@ func TestFirstSet2(t *testing.T) {
         t.Error("failed to calculate first set of E\n")
     }
 }
+
+func TestEBNF1(t *testing.T) {
+    text := `expr : INT | ID | expr;`
+    tokens := []int{2, 9, 1, 5, 1, 5, 2, 10, -1}
```
Comment (anzdaddy) [inline sysl2/Grammar1_test.go]: Magic numbers?

File: sysl2/ebnfParser.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        a = makeStringAtom(tokText)
+    case 2:
+        a = makeRuleNameAtom(symbolTerm(atom[0]).tok.text)
+    default:
+        panic("not implemented yet.")
+    }
+    return a
+}
+
+func symbolTerm(item interface{}) symbol {
```
Comment (anzdaddy) [inline sysl2/ebnfParser.go]: Seems overly verbose. Why not just type-assert inline?

File: sysl2/ebnfParser.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        return c, seq
+    }
+    return -1, nil
+}
+
+// ruleSeq returns Rule.Choice.Sequence
+func ruleSeq(item interface{}, rulename string) (int, []interface{}) {
+    rule, ok := item.(map[string]map[int][]interface{})
+
+    if ok {
```
Comment (anzdaddy) [inline sysl2/ebnfParser.go]: Fold into previous line.

File: sysl2/ebnfParser.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        return getChoice(rule[rulename])
+    }
+    return -1, nil
+}
+
+func buildSequence(s0 []interface{}) *sysl.Sequence {
+    terms := make([]*sysl.Term, 0)
+
+    if s0 != nil {
```
Comment (anzdaddy) [inline sysl2/ebnfParser.go]: Not enough. Could be non-nil but empty. Use `len(s0) == 0` instead.

File: sysl2/ebnfParser.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            t := seq.([]interface{})[0]
+            if t == nil {
+                break
+            }
+            tt := t.(map[int][]interface{})[0]
+
+            _, s0 = ruleSeq(tt[1], "seq")
+        } else {
+            _, s0 = ruleSeq(seq, "seq")
+        }
```
Comment (anzdaddy) [inline sysl2/ebnfParser.go]: Possible alternate form:

```go
    for option, seq := range choice {
        if option > 0 {
            ⋮
            seq = tt[1]
        }
        _, s0 := ruleSeq(seq, "seq")
```

File: sysl2/grammarParser.go [hunk 8]
```diff
@@ -79,8 +89,143 @@ func makeParser(g *sysl.Grammar, text string) *parser {
...
+                // nil == epsilon
+                fmt.Println("matched nil")
+                result = true
+                continue
+            }
+            minCount, maxCount := getMinMaxCount(t)
+            matchCount := 0
+            res := false
+            subTree := make([]interface{}, 0)
+            singleTerm := minCount == maxCount && minCount == 1
```
Comment (anzdaddy) [inline sysl2/grammarParser.go]: Confusing. More obvious: `minCount == 1 && maxCount == 1`.

File: sysl2/grammarParser.go [hunk 9]
```diff
@@ -79,8 +89,143 @@ func makeParser(g *sysl.Grammar, text string) *parser {
...
+    case *sysl.Quantifier_OnePlus: // +
+        return 1, math.MaxInt32
+        // case *sysl.Quantifier_Separator:
+    }
+    return 1, 1
+}
+
+func (p *parser) parse(g *sysl.Grammar, input int, val interface{}) (bool, int, []interface{}) {
+    if input == len(*(p.tokens)) {
+        fmt.Println("got input length zero!!!")
```
Comment (anzdaddy) [inline sysl2/grammarParser.go]: Lots of stdout logging. Should be something more pluggable.

File: sysl2/sampleGrammars.go [hunk 10]
```diff
@@ -18,7 +18,7 @@ func makeQuantifierOnePlus() *sysl.Quantifier {
     return &sysl.Quantifier{Union: &sysl.Quantifier_OnePlus{}}
 }
 
-func makeTerm(str string) *sysl.Term {
+func makeStringTerm(str string) *sysl.Term {
     return &sysl.Term{Atom: &sysl.Atom{Union: &sysl.Atom_Regexp{Regexp: str}}, Quantifier: nil}
```
Comment (anzdaddy) [inline sysl2/sampleGrammars.go]: Line too long

**Pull request closed without merging**